src/Icicle/Core/Exp/Simp.hs

```diff
@@ -62,125 +60,27 @@ simpX isValue = go
-      M.PrimArith M.PrimArithMul
-        -> arith2 pp args (*)
-      M.PrimArith M.PrimArithNegate
-        | [(_, VInt x)] <- args
-        -> let t = functionReturns (M.typeOfPrim pp)
-           in  return $ XValue t (VInt (-x))
-        | otherwise -> Nothing
-
-      -- * predicates on integers
-      M.PrimRelation M.PrimRelationGt IntT
-        -> int2 args (>)
-      M.PrimRelation M.PrimRelationGe IntT
-        -> int2 args (>=)
-      M.PrimRelation M.PrimRelationLt IntT
-        -> int2 args (<)
-      M.PrimRelation M.PrimRelationLe IntT
-        -> int2 args (<)
-      M.PrimRelation M.PrimRelationEq IntT
-        -> int2 args (==)
-      M.PrimRelation M.PrimRelationNe IntT
-        -> int2 args (/=)
-
-      -- * predicates on bools
-      M.PrimRelation M.PrimRelationGt BoolT
-        -> bool2 args (>)
-      M.PrimRelation M.PrimRelationGe BoolT
-        -> bool2 args (>=)
-      M.PrimRelation M.PrimRelationLt BoolT
-        -> bool2 args (<)
-      M.PrimRelation M.PrimRelationLe BoolT
-        -> bool2 args (<)
-      M.PrimRelation M.PrimRelationEq BoolT
-        -> bool2 args (==)
-      M.PrimRelation M.PrimRelationNe BoolT
-        -> bool2 args (/=)
-
-      M.PrimRelation _ _ -> Nothing
-
-      -- * logical
-      M.PrimLogical M.PrimLogicalAnd
-        -> bool2 args (&&)
-      M.PrimLogical M.PrimLogicalOr
-        -> bool2 args (||)
-      M.PrimLogical M.PrimLogicalNot
-        -> bool1 args not
-
-      -- * constructors
-      M.PrimConst    (M.PrimConstPair _ _)
-       | [(ta,va),(tb,vb)] <- args
-       -> return $ XValue (PairT ta tb) (VPair va vb)
-       | otherwise
-       -> Nothing
-
-      M.PrimConst    (M.PrimConstSome _)
-       | [(ta,va)] <- args
-       -> return $ XValue (OptionT ta) (VSome va)
-       | otherwise
-       -> Nothing
-
-      -- * leaves datetime alone
-      M.PrimDateTime _ -> Nothing
-
-      M.PrimPair    (M.PrimPairFst ta _)
-       | [(_, VPair va _)] <- args
-       -> return $ XValue ta va
-       | otherwise
-       -> Nothing
-
-      M.PrimPair    (M.PrimPairSnd _ tb)
-       | [(_, VPair _ vb)] <- args
-       -> return $ XValue tb vb
-       | otherwise
-       -> Nothing
-
-
-    bool1 args f
-      | [(_, VBool x)] <- args
-      = return $ XValue BoolT (VBool (f x))
-      | otherwise = Nothing
-
-    bool2 args f
-      | [(_, VBool x), (_, VBool y)] <- args
-      = return $ XValue BoolT (VBool (f x y))
-      | otherwise = Nothing
-
-    int2 args f
-      | [(_, VInt x), (_, VInt y)] <- args
-      = return $ XValue BoolT (VBool (f x y))
-      | otherwise = Nothing
-
-    arith2 pp args f
-      | [(_, VInt x), (_, VInt y)] <- args
-      = let t = functionReturns (M.typeOfPrim pp)
-            v = f x y
-        in  return $ XValue t (VInt v)
-      | otherwise = Nothing
-
-
-takeValue :: Exp n p -> Maybe (ValType, BaseValue)
-takeValue (XValue a b) = Just (a, b)
-takeValue _            = Nothing
+simpP :: Ord n => Prim -> [Value n Prim] -> Maybe (C.Exp n)
```

Comment:
Actually the main advantage that I see of doing "real partial evaluation" rather than calling evalPrim would be that it could rewrite the above Array_map. Maybe evalPrim could be generalised to support both cases, I'm not sure